# 📊 01 Data Understanding & Validation

## Project: Bank Customer Churn Prediction

This notebook focuses on understanding and validating the raw bank customer churn dataset before performing exploratory data analysis, feature engineering, or machine learning.

The goal of this step is to make sure the dataset is clean, reliable, and ready for deeper analysis.

---

## 🎯 Purpose of This Notebook

In this notebook, we will answer the following questions:

| Area | Validation Question |
|---|---|
| Dataset Structure | What data do we have? |
| Unit of Analysis | What does each row represent? |
| Data Dictionary | What does each column mean? |
| Missing Values | Are there any missing values? |
| Duplicate Records | Are there duplicate customers or rows? |
| Target Variable | Is the churn column valid? |
| Categorical Columns | Are categorical values clean and consistent? |
| Binary Columns | Do binary columns contain only valid 0/1 values? |
| Numerical Columns | Are numerical values realistic? |

---

## ✅ Expected Outcome

By the end of this notebook, we should know whether the dataset is suitable for the next stage of the project.

If the data passes the validation checks, we can confidently move forward to exploratory data analysis.

> **Data Scientist Mindset:**  
> Before building models or dashboards, we must first understand the data, validate its quality, and confirm that the target variable is correctly defined.

## 1. Load Dataset

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv(r"C:\Users\amish\OneDrive\Desktop\portfolio\fintech-customer-churn-retention\data\raw\Bank Customer Churn Prediction.csv")

df.head()

,customer_id,credit_score,country,gender,age,tenure,balance,products_number,credit_card,active_member,estimated_salary,churn
0,15634602,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,15647311,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,15619304,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,15701354,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,15737888,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


### Interpretation

The dataset was loaded successfully.

The first five rows show that each record represents **one bank customer**. The dataset includes both customer profile information and banking-related variables.

**Customer profile variables include:**

- `country`
- `gender`
- `age`

**Banking-related variables include:**

- `credit_score`
- `tenure`
- `balance`
- `products_number`
- `credit_card`
- `active_member`
- `estimated_salary`

The target variable is `churn`, where:

| Value | Meaning |
|---|---|
| `0` | Customer stayed |
| `1` | Customer churned |

## 2. Dataset Shape

The dataset shape is checked to understand how many customer records and columns are available.

In [3]:
df.shape

(10000, 12)

### Interpretation

The dataset contains **10,000 customer records** and **12 columns**.

Each row represents one bank customer, and each column represents customer profile information, banking behavior, or the churn target variable.

This confirms that the dataset has enough records for customer churn analysis and future machine learning work.

## 3. Column Review

Column names are reviewed to understand what variables are available in the dataset and whether the names are clear, consistent, and easy to use.

In [5]:
df.columns

Index(['customer_id', 'credit_score', 'country', 'gender', 'age', 'tenure',
       'balance', 'products_number', 'credit_card', 'active_member',
       'estimated_salary', 'churn'],
      dtype='object')

### Interpretation

The dataset contains **12 columns**.

The column names are clean, readable, and consistently formatted using lowercase words with underscores.

The dataset includes:

- customer identification: `customer_id`
- customer profile variables: `country`, `gender`, `age`
- banking-related variables: `credit_score`, `tenure`, `balance`, `products_number`, `credit_card`, `active_member`, `estimated_salary`
- target variable: `churn`

No column renaming is required at this stage.

## 4. Data Type Validation

Data types are checked to confirm that each column is stored in the correct format before analysis.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customer_id       10000 non-null  int64  
 1   credit_score      10000 non-null  int64  
 2   country           10000 non-null  object 
 3   gender            10000 non-null  object 
 4   age               10000 non-null  int64  
 5   tenure            10000 non-null  int64  
 6   balance           10000 non-null  float64
 7   products_number   10000 non-null  int64  
 8   credit_card       10000 non-null  int64  
 9   active_member     10000 non-null  int64  
 10  estimated_salary  10000 non-null  float64
 11  churn             10000 non-null  int64  
dtypes: float64(2), int64(8), object(2)
memory usage: 937.6+ KB


### Interpretation

The dataset contains **10,000 records** and **12 columns**.

The data types appear appropriate for this dataset:

- `customer_id` is stored as an integer identifier.
- `country` and `gender` are stored as object/categorical variables.
- `balance` and `estimated_salary` are stored as numerical float variables.
- `credit_score`, `age`, `tenure`, `products_number`, `credit_card`, `active_member`, and `churn` are stored as integer variables.

No incorrect data types were found at this stage.

## 5. Missing Value Validation

Missing values are checked to identify whether any columns have incomplete data that may require cleaning, imputation, or removal.

In [7]:
missing_report = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percent": (df.isnull().sum() / len(df)) * 100
})

missing_report

,missing_count,missing_percent
customer_id,0,0.0
credit_score,0,0.0
country,0,0.0
gender,0,0.0
age,0,0.0
tenure,0,0.0
balance,0,0.0
products_number,0,0.0
credit_card,0,0.0
active_member,0,0.0


### Interpretation

No missing values were found in the dataset.

This means all 10,000 customer records have complete information across the available columns. Since there are no missing values, no imputation or row removal is required at this stage.

## 6. Duplicate Record Validation

Duplicate records are checked to make sure the same customer record is not repeated in the dataset.

In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
df.duplicated(subset=["customer_id"]).sum()

np.int64(0)

### Interpretation

No duplicate rows were found in the dataset.

No duplicate `customer_id` values were found, which means each customer appears only once in the dataset.

This confirms that the dataset does not contain repeated customer records.

## 7. Target Variable Validation

The target variable is checked to confirm that it is correctly defined and contains valid values.

In this dataset, the target variable is `churn`.

In [11]:
df["churn"].value_counts()

churn
0    7963
1    2037
Name: count, dtype: int64

In [12]:
df["churn"].value_counts(normalize=True) * 100

churn
0    79.63
1    20.37
Name: proportion, dtype: float64

### Interpretation

The target variable `churn` contains only valid binary values:

| Churn Value | Meaning |
|---|---|
| `0` | Customer stayed |
| `1` | Customer churned |

The dataset contains **7,963 retained customers** and **2,037 churned customers**.

This means approximately **79.63%** of customers stayed and **20.37%** of customers churned.

The dataset has a moderate class imbalance because retained customers are more common than churned customers. This is normal for churn problems and should be considered later during model evaluation.

## 8. Categorical Variable Validation

Categorical variables are checked to identify inconsistent labels, spelling issues, unexpected categories, or formatting problems.

In this dataset, the main categorical variables are:

- `country`
- `gender`

In [13]:
categorical_cols = ["country", "gender"]

for col in categorical_cols:
    print(f"Column: {col}")
    print(df[col].value_counts())
    print("-" * 40)

Column: country
country
France     5014
Germany    2509
Spain      2477
Name: count, dtype: int64
----------------------------------------
Column: gender
gender
Male      5457
Female    4543
Name: count, dtype: int64
----------------------------------------


### Interpretation

The categorical variables are clean and consistent.

The `country` column contains three categories:

- France
- Germany
- Spain

The `gender` column contains two categories:

- Male
- Female

No spelling issues, inconsistent labels, missing categories, or unexpected values were found in these categorical columns.

## 9. Binary Variable Validation

Binary variables are checked to confirm that they contain only valid `0` and `1` values.

In this dataset, the binary variables are:

- `credit_card`
- `active_member`
- `churn`

In [14]:
binary_cols = ["credit_card", "active_member", "churn"]

for col in binary_cols:
    print(f"Column: {col}")
    print(df[col].value_counts())
    print("-" * 40)

Column: credit_card
credit_card
1    7055
0    2945
Name: count, dtype: int64
----------------------------------------
Column: active_member
active_member
1    5151
0    4849
Name: count, dtype: int64
----------------------------------------
Column: churn
churn
0    7963
1    2037
Name: count, dtype: int64
----------------------------------------


### Interpretation

The binary variables contain only valid `0` and `1` values.

- `credit_card` indicates whether the customer has a credit card.
- `active_member` indicates whether the customer is an active member.
- `churn` indicates whether the customer stayed or left the bank.

No invalid binary values were found in these columns.

## 10. Numerical Range Validation

Numerical variables are checked to identify unrealistic values, negative values, or unusual ranges that may require further investigation.

This helps confirm whether the numerical columns are reasonable for customer churn analysis.

In [15]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
customer_id,10000.0,1.569094e+07,71936.186123,15565701.00,15628528.25,1.569074e+07,1.575323e+07,15815690.00
credit_score,10000.0,6.505288e+02,96.653299,350.00,584.00,6.520000e+02,7.180000e+02,850.00
age,10000.0,3.892180e+01,10.487806,18.00,32.00,3.700000e+01,4.400000e+01,92.00
tenure,10000.0,5.012800e+00,2.892174,0.00,3.00,5.000000e+00,7.000000e+00,10.00
balance,10000.0,7.648589e+04,62397.405202,0.00,0.00,9.719854e+04,1.276442e+05,250898.09
products_number,10000.0,1.530200e+00,0.581654,1.00,1.00,1.000000e+00,2.000000e+00,4.00
credit_card,10000.0,7.055000e-01,0.455840,0.00,0.00,1.000000e+00,1.000000e+00,1.00
active_member,10000.0,5.151000e-01,0.499797,0.00,0.00,1.000000e+00,1.000000e+00,1.00
estimated_salary,10000.0,1.000902e+05,57510.492818,11.58,51002.11,1.001939e+05,1.493882e+05,199992.48
churn,10000.0,2.037000e-01,0.402769,0.00,0.00,0.000000e+00,0.000000e+00,1.00


In [16]:
print("Credit score range:", df["credit_score"].min(), "-", df["credit_score"].max())
print("Age range:", df["age"].min(), "-", df["age"].max())
print("Tenure range:", df["tenure"].min(), "-", df["tenure"].max())
print("Balance range:", df["balance"].min(), "-", df["balance"].max())
print("Products number range:", df["products_number"].min(), "-", df["products_number"].max())
print("Estimated salary range:", df["estimated_salary"].min(), "-", df["estimated_salary"].max())

Credit score range: 350 - 850
Age range: 18 - 92
Tenure range: 0 - 10
Balance range: 0.0 - 250898.09
Products number range: 1 - 4
Estimated salary range: 11.58 - 199992.48


### Interpretation

The numerical columns have reasonable ranges for this dataset.

- `credit_score` ranges from 350 to 850.
- `age` ranges from 18 to 92.
- `tenure` ranges from 0 to 10 years.
- `balance` values are non-negative.
- `products_number` ranges from 1 to 4.
- `estimated_salary` values are non-negative.

No clearly invalid numerical values, such as negative balances, negative salaries, or impossible ages, were found during the initial validation.

## 11. Feature Type Grouping

Columns are grouped by feature type to make the dataset easier to understand and prepare for future analysis.

This helps separate identifier columns, target variables, categorical variables, binary variables, and numerical variables.

In [17]:
id_col = ["customer_id"]

target_col = ["churn"]

categorical_cols = ["country", "gender"]

binary_cols = ["credit_card", "active_member"]

numeric_cols = [
    "credit_score",
    "age",
    "tenure",
    "balance",
    "products_number",
    "estimated_salary"
]

print("ID column:", id_col)
print("Target column:", target_col)
print("Categorical columns:", categorical_cols)
print("Binary columns:", binary_cols)
print("Numerical columns:", numeric_cols)

ID column: ['customer_id']
Target column: ['churn']
Categorical columns: ['country', 'gender']
Binary columns: ['credit_card', 'active_member']
Numerical columns: ['credit_score', 'age', 'tenure', 'balance', 'products_number', 'estimated_salary']


### Interpretation

The dataset columns were grouped by their role in the project.

- `customer_id` is an identifier column and should not be used as a predictive feature.
- `churn` is the target variable.
- `country` and `gender` are categorical variables.
- `credit_card` and `active_member` are binary variables.
- `credit_score`, `age`, `tenure`, `balance`, `products_number`, and `estimated_salary` are numerical variables.

This grouping will help organize future steps such as exploratory data analysis, preprocessing, and machine learning.

## 12. Data Understanding & Validation Summary

The dataset contains **10,000 customer records** and **12 columns**. Each row represents one bank customer.

The target variable is `churn`, where:

| Churn Value | Meaning |
|---|---|
| `0` | Customer stayed |
| `1` | Customer churned |

The dataset passed the initial validation checks:

- No missing values were found.
- No duplicate rows were found.
- No duplicate customer IDs were found.
- The target variable contains valid binary values.
- Categorical columns are clean and consistent.
- Binary columns contain only valid `0` and `1` values.
- Numerical columns have reasonable ranges.
- The dataset has moderate class imbalance, with more retained customers than churned customers.

The `customer_id` column will be kept for identification purposes but excluded from future machine learning models because it does not represent customer behavior.

Based on these checks, the dataset is clean, reliable, and ready for deeper analysis.